MdeModulePkg/Bus/Pci/PciBusDxe/PciDoe.c

```diff
@@ -0,0 +1,511 @@
+
+**/
+EFI_STATUS
+EFIAPI
+PciIoDoeTypeCheck (
+  IN EDKII_PCI_DOE_PROTOCOL  *This,
+  IN     UINTN               VendorId,
+  IN     UINTN               Type
+  )
+{
+  EFI_STATUS                        Status;
+  PCI_IO_DEVICE                     *PciIoDevice;
+  PCI_EXPRESS_DOE_DISCOVERY_PACKET  Packet;
+  UINT8                             Index;
+  UINTN                             Count;
+  UINT16                            ResponseVid, ResponseType;
+  UINT32                            Response[2];
+
+  PciIoDevice = PCI_IO_DEVICE_FROM_PCI_DOE_THIS (This);
+
+  Status = DoeWaitForNotBusy (PciIoDevice);
+  if (EFI_ERROR (Status)) {
+    return EFI_UNSUPPORTED;
+  }
+
+  Packet.Header1 = (PCI_DOE_PROTOCOL_DISCOVERY << DOE_HEADER1_OFST_TYPE) | (PCI_VENDOR_ID_PCI_SIG << DOE_HEADER1_OFST_VID);
+  Packet.Header2 = 3 << DOE_HEADER2_OFST_LEN;
+
+  Index = 0;
+
+  do {
+    Packet.Dw0 = Index;
+
+    DEBUG ((DEBUG_INFO, "Sending DOE Discovery packet\n"));
+
+    // Send the DOE discovery packet
+    Count = 3 * 4;
+
+    Status = PciIoDoeSend (This, Count, &Packet);
+    if (EFI_ERROR (Status)) {
+      return EFI_UNSUPPORTED;
+    }
+
+    // Set DOE Go bit
+    Status = DoeSetGoBit (PciIoDevice);
+    if (EFI_ERROR (Status)) {
+      return EFI_UNSUPPORTED;
+    }
+
+    // Wait for a response
+    Status = DoeWaitForStatusDor (PciIoDevice);
+    if (EFI_ERROR (Status)) {
+      return EFI_UNSUPPORTED;
+    }
+
+    // Read response
+    Count  = 2 * 4;
+    Status = PciIoDoeRecieve (This, &Count, &Response);
+    if (EFI_ERROR (Status)) {
+      return EFI_UNSUPPORTED;
+    }
+
+    DEBUG ((DEBUG_INFO, "Response is: 0x%x/0x%x\n", Response[0], Response[1]));
+
+    // Bits 0:15
+    ResponseVid = Response[0] & 0xFFFF;
+    // Bits 16:23
+    ResponseType = (Response[0] >> 16) & 0xFF;
+    // Bits 24:31
+    Index = Response[0] >> 24;
+
+    if ((VendorId == ResponseVid) && (Type == ResponseType)) {
+      return EFI_SUCCESS;
+    }
+
+    // Check if Index matches
+  } while (Index > 0);
+
+  return RETURN_UNSUPPORTED;
+}
+
+//
+// Pci Io DOE Interface
+//
+EDKII_PCI_DOE_PROTOCOL  mPciIoDoeInterface = {
+  PciIoDoeTypeCheck,
+  PciIoDoeSend,
+  PciIoDoeRecieve,
+};
+
+/**
+  Probe if the PCIe device supports DOE and if it does populate the
+  EDKII_PCI_DOE_PROTOCOL instance.
+
+  @param PciIoDevice      PCI device instance.
+
+  @retval EFI_UNSUPPORTED PCI Device does not support DOE.
+  @retval EFI_SUCCESS     PCI Device does support DOE.
+
+**/
+EFI_STATUS
+ProbeDoeSupport (
+  IN PCI_IO_DEVICE  *PciIoDevice
+  )
+{
+  EFI_STATUS  Status;
+
+  Status = LocatePciExpressCapabilityRegBlock (
+             PciIoDevice,
+             EFI_PCIE_CAPABILITY_ID_DOE,
+             &PciIoDevice->DoeCapabilityOffset,
+             NULL
+             );
+  if (EFI_ERROR (Status)) {
+    return EFI_UNSUPPORTED;
+  }
+
+  DEBUG ((DEBUG_INFO, "Found DOE support!\n"));
+
+  CopyMem (&PciIoDevice->PciDoe, &mPciIoDoeInterface, sizeof (EDKII_PCI_DOE_PROTOCOL));

```

Comment:
I am confused.  If the Doe services are implemented in the PCI Bus Driver and are part of the PCI I/O Device structure that is internal to the PCI Bus Driver, then a protocol is not needed at all.  

A protocol should only be defined if one component produced the protocol services with gBS->InstallProtocolInterface() to register services in the handle database and is consumed by a different component that looks it up using gBS->LocateProtocol() or gBS->HandleProtocol(), or gBS->OpenProtocol().

Reply:
I'm not super familiar with the EDK2 code base, but DOE is a protocol used to send messages to a PCIe device. Which is why I feel it should be exposed as a protocol in EDK2.

The idea being that multiple consumers can use the protocol with `gBS->LocateProtocol()`, `gBS->HandleProtocol()` or `gBS->OpenProtocol()` to implement functionality on top of DOE, such as SPDM

Reply:
A communications protocol is not the same as a UEFI Firmware Protocol.  If the PCI Bus Driver is the only module that needs to use the communications protocol to interact with a PCI controller, then it can be internal APIs to the PCI Bus Driver.  Just like PCI Enumeration.